flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/BinaryToBinaryCastRule.java

```diff
@@ -43,37 +43,38 @@ private BinaryToBinaryCastRule() {
                         .build());
     }
 
-    /* Example generated code for BINARY(2):
-
-    // legacy behavior
-    ((byte[])(inputValue))
-
-    // new behavior
-    ((((byte[])(inputValue)).length == 2) ? (((byte[])(inputValue))) : (java.util.Arrays.copyOf(((byte[])(inputValue)), 2)))
-
-    */
-
+    /**
+     * Generates code for casting between BINARY and VARBINARY types.
+     *
+     * <p>For VARBINARY targets: preserves original length if it fits within the target constraint,
+     * otherwise truncates to target length.
+     *
+     * <p>For BINARY targets: pads shorter inputs to exact target length, truncates longer inputs.
+     *
+     * <p>Example generated code for {@code CAST(input AS VARBINARY(4))}:
+     *
+     * <pre>
+     * ((input.length <= 4) ? input : java.util.Arrays.copyOf(input, 4))
```

Comment:
the down side of this approach is that we always perform a length check during even if it is not necessary